include/nbl/asset/filters/CBlitUtilities.h

```diff
@@ -24,50 +27,67 @@ class IBlitUtilities
-	static inline size_t getScaledKernelPhasedLUTSize(const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent, const asset::IImage::E_TYPE inImageType,
-		const KernelX& kernelX, const KernelY& kernelY, const KernelZ& kernelZ)
+	using convolution_kernels_t = std::tuple<
+		CFloatingPointSeparableImageFilterKernel<CConvolutionWeightFunction1D<ReconstructionFunctionX, ResamplingFunctionX>>,
+		CFloatingPointSeparableImageFilterKernel<CConvolutionWeightFunction1D<ReconstructionFunctionY, ResamplingFunctionY>>,
+		CFloatingPointSeparableImageFilterKernel<CConvolutionWeightFunction1D<ReconstructionFunctionZ, ResamplingFunctionZ>>>;
+
+	using convolution_kernel_x_t = std::tuple_element_t<0, convolution_kernels_t>;
+	using convolution_kernel_y_t = std::tuple_element_t<1, convolution_kernels_t>;
+	using convolution_kernel_z_t = std::tuple_element_t<2, convolution_kernels_t>;
+
+	using value_type = convolution_kernel_x_t::value_type;
+	static_assert(std::is_same_v<value_type, convolution_kernel_y_t::value_type> && std::is_same_v<value_type, convolution_kernel_z_t::value_type>);
+	static inline constexpr uint32_t MaxChannels = convolution_kernel_x_t::MaxChannels;
+
+	template <typename LutDataType>
+	static inline size_t getScaledKernelPhasedLUTSize(
+		const core::vectorSIMDu32&		inExtent,
+		const core::vectorSIMDu32&		outExtent,
+		const asset::IImage::E_TYPE		inImageType,
+		const convolution_kernels_t&	kernels)
 	{
-		const auto scaledKernelX = constructScaledKernel(kernelX, inExtent, outExtent);
-		const auto scaledKernelY = constructScaledKernel(kernelY, inExtent, outExtent);
-		const auto scaledKernelZ = constructScaledKernel(kernelZ, inExtent, outExtent);
+		const auto windowSize = getWindowSize(inImageType, kernels);
+		return getScaledKernelPhasedLUTSize<LutDataType>(inExtent, outExtent, inImageType, windowSize);
+	}
 
+	template <typename LutDataType>
+	static inline size_t getScaledKernelPhasedLUTSize(
+		const core::vectorSIMDu32&		inExtent,
+		const core::vectorSIMDu32&		outExtent,
+		const asset::IImage::E_TYPE		inImageType,
+		const core::vectorSIMDi32&		windowSize)
+	{
 		const auto phaseCount = getPhaseCount(inExtent, outExtent, inImageType);
-
-		return ((phaseCount[0] * scaledKernelX.getWindowSize().x) + (phaseCount[1] * scaledKernelY.getWindowSize().y) + (phaseCount[2] * scaledKernelZ.getWindowSize().z)) * sizeof(lut_value_type) * MaxChannels;
+		return ((phaseCount.x * windowSize.x) + (phaseCount.y * windowSize.y) + (phaseCount.z * windowSize.z)) * sizeof(LutDataType) * MaxChannels;
 	}
 
-	template <typename lut_value_type = typename KernelX::value_type>
-	static bool computeScaledKernelPhasedLUT(void* outKernelWeights, const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent, const asset::IImage::E_TYPE inImageType,
-		const KernelX& kernelX, const KernelY& kernelY, const KernelZ& kernelZ)
+	template <typename LutDataType>
+	static bool computeScaledKernelPhasedLUT(
+		void*							outKernelWeights,
```

Comment:
The method needs a `const double normalizeWeightsTo=1.0` argument

You need to compute the normalization factor if `!core::isnan(normalizeWeightsTo)` which would be the `normalizeWeightsTo` divided by the sum of all weights for a given axis & phase.

This is why:
https://discord.com/channels/593902898015109131/1043939034776473710/1062723347835793428
https://discord.com/channels/593902898015109131/1043939034776473710/1062724151699308655
https://discord.com/channels/593902898015109131/1043939034776473710/1067676795391967323

Basically even having a reconstruction filter won't help you against value fluctuations (ringing) like these:
![image](https://github.com/Devsh-Graphics-Programming/Nabla/assets/6894321/ef5b02b4-f6df-497d-8959-63d55adf2446)

Why?

Imagine you're using the convolution of two gaussians or two sinc filters for your kernel.

In the gaussian case you'll have a variance thats a sum of the two input gaussians, in the sinc you'll be left with the wider one.

Now imagine you're upscaling a constant value image, so you have many many phases and a really wide kernel in input coordinates.

**Even though the kernel resulting from the convolution integrates to a constant, you won't get a constant image!**

Because there's a finite amount of input pixels within the kernel's support, its not a real "full" integration, and therefore its like doing a low sample count Riemann Sum with slightly shifted (by the phase) sample points for each output pixel.

Since Riemann Sum != The True Integral if the sample count does not increase to infinity, you'll get under or oversampling.

Reply:
Btw you can't quantize/compress the weights to `float` or `float16_t` before you're done normalizing them, so you should always require `outKernelWeights` be `sizeof(double)` sized

Reply:
Resolved